// @mantine/form/esm/hooks/use-form-list/use-form-list.mjs
"use client";
import { changeErrorIndices } from "../../lists/change-error-indices.mjs";
import { reorderErrors } from "../../lists/reorder-errors.mjs";
import { reorderPath } from "../../paths/reorder-path.mjs";
import { insertPath } from "../../paths/insert-path.mjs";
import { removePath } from "../../paths/remove-path.mjs";
import { replacePath } from "../../paths/replace-path.mjs";
import { useCallback } from "react";
//#region packages/@mantine/form/src/hooks/use-form-list/use-form-list.ts
function useFormList({ $values, $errors, $status }) {
	return {
		reorderListItem: useCallback((path, payload) => {
			$status.clearFieldDirty(path);
			$errors.setErrors((errs) => reorderErrors(path, payload, errs));
			$values.setValues({
				values: reorderPath(path, payload, $values.refValues.current),
				updateState: true
			});
		}, []),
		removeListItem: useCallback((path, index) => {
			$status.clearFieldDirty(path);
			$errors.setErrors((errs) => changeErrorIndices(path, index, errs, -1));
			$values.setValues({
				values: removePath(path, index, $values.refValues.current),
				updateState: true
			});
		}, []),
		insertListItem: useCallback((path, item, index) => {
			$status.clearFieldDirty(path);
			$errors.setErrors((errs) => changeErrorIndices(path, index, errs, 1));
			$values.setValues({
				values: insertPath(path, item, index, $values.refValues.current),
				updateState: true
			});
		}, []),
		replaceListItem: useCallback((path, index, item) => {
			$status.clearFieldDirty(path);
			$values.setValues({
				values: replacePath(path, item, index, $values.refValues.current),
				updateState: true
			});
		}, [])
	};
}
//#endregion
export { useFormList };
