// @mantine/core/cjs/components/Tree/use-tree-node-drag-drop.cjs
"use client";
const require_get_children_nodes_values = require("./get-children-nodes-values/get-children-nodes-values.cjs");
//#region packages/@mantine/core/src/components/Tree/use-tree-node-drag-drop.ts
function isDescendantOf(data, ancestorValue, descendantValue) {
	const ancestor = require_get_children_nodes_values.findTreeNode(ancestorValue, data);
	if (!ancestor || !ancestor.children) return false;
	function check(nodes) {
		for (const node of nodes) {
			if (node.value === descendantValue) return true;
			if (node.children && check(node.children)) return true;
		}
		return false;
	}
	return check(ancestor.children);
}
function getDragDropPosition(event, element, hasChildren) {
	const rect = element.getBoundingClientRect();
	const y = event.clientY - rect.top;
	const height = rect.height;
	if (hasChildren) {
		if (y < height * .25) return "before";
		if (y > height * .75) return "after";
		return "inside";
	}
	if (y < height * .5) return "before";
	return "after";
}
const EMPTY_DRAG_PROPS = {};
function useTreeNodeDragDrop({ nodeValue, hasChildren, data, onDragDrop, dragStateRef }) {
	if (!onDragDrop) return EMPTY_DRAG_PROPS;
	const handleDragStart = (event) => {
		event.stopPropagation();
		event.dataTransfer.effectAllowed = "move";
		event.dataTransfer.setData("text/plain", nodeValue);
		dragStateRef.current.draggedValue = nodeValue;
		const target = event.currentTarget;
		requestAnimationFrame(() => {
			target.setAttribute("data-dragging", "true");
		});
	};
	const handleDragOver = (event) => {
		const draggedValue = dragStateRef.current.draggedValue;
		if (!draggedValue || draggedValue === nodeValue) return;
		if (isDescendantOf(data, draggedValue, nodeValue)) return;
		event.preventDefault();
		event.stopPropagation();
		event.dataTransfer.dropEffect = "move";
		const target = event.currentTarget;
		const position = getDragDropPosition(event, target, hasChildren);
		const prevTarget = dragStateRef.current.currentDropTarget;
		if (prevTarget && prevTarget !== target) prevTarget.removeAttribute("data-drag-over");
		target.setAttribute("data-drag-over", position);
		dragStateRef.current.currentDropTarget = target;
	};
	const handleDragLeave = (event) => {
		const target = event.currentTarget;
		const related = event.relatedTarget;
		if (related && target.contains(related)) return;
		target.removeAttribute("data-drag-over");
		if (dragStateRef.current.currentDropTarget === target) dragStateRef.current.currentDropTarget = null;
	};
	const handleDrop = (event) => {
		event.preventDefault();
		event.stopPropagation();
		const target = event.currentTarget;
		const position = target.getAttribute("data-drag-over");
		target.removeAttribute("data-drag-over");
		const draggedValue = dragStateRef.current.draggedValue;
		if (draggedValue && position && draggedValue !== nodeValue) onDragDrop({
			draggedNode: draggedValue,
			targetNode: nodeValue,
			position
		});
		dragStateRef.current.draggedValue = null;
		dragStateRef.current.currentDropTarget = null;
	};
	const handleDragEnd = (event) => {
		event.currentTarget.removeAttribute("data-dragging");
		const prevTarget = dragStateRef.current.currentDropTarget;
		if (prevTarget) prevTarget.removeAttribute("data-drag-over");
		dragStateRef.current.draggedValue = null;
		dragStateRef.current.currentDropTarget = null;
	};
	return {
		draggable: true,
		onDragStart: handleDragStart,
		onDragOver: handleDragOver,
		onDragLeave: handleDragLeave,
		onDrop: handleDrop,
		onDragEnd: handleDragEnd
	};
}
//#endregion
exports.useTreeNodeDragDrop = useTreeNodeDragDrop;
